File: src/main/java/catwalks/util/meta/FloatProp.java
package catwalks.util.meta;

/**
 * Created by TheCodeWarrior
 */
public class FloatProp extends MetaValueProperty<Float> {
	
	protected IntProp intProp;
	protected float coefficient;
	
	public FloatProp(int index, int bits) {
		intProp = new IntProp(index, bits);
		coefficient = (float)Math.pow(2, bits/2);
	}
	
	@Override
	public boolean addBits(int index, int bits) {
		intProp.addBits(index, bits);
		return true;
	}
	
	@Override
	public void set(MetaStorage storage, Float value) {
		int intVal = (int)( value * coefficient );
		
		intProp.set(storage, intVal);
		storage.notifyIfDirty();
	}
	
	@Override
	public Float get(MetaStorage storage) {
		int intVal = intProp.get(storage);
		return intVal / coefficient;
	}
}

File: src/main/java/catwalks/util/meta/FloatProp.kt
package catwalks.util.meta

/**
 * Created by TheCodeWarrior
 */
class FloatProp(index: Int, bits: Int) : MetaValueProperty<Float>() {

    protected var intProp: IntProp
    protected var coefficient: Float = 0.toFloat()

    init {
        intProp = IntProp(index, bits)
        coefficient = Math.pow(2.0, (bits / 2).toDouble()).toFloat()
    }

    override fun addBits(index: Int, bits: Int): Boolean {
        intProp.addBits(index, bits)
        return true
    }

    override fun set(storage: MetaStorage, value: Float) {
        val intVal = (value * coefficient).toInt()

        intProp.set(storage, intVal)
        storage.notifyIfDirty()
    }

    override fun get(storage: MetaStorage): Float {
        val intVal = intProp.get(storage)
        return intVal / coefficient
    }
}
